url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+        recentFunctionValue = math.min(fit.raw).val;
+        recentMaxFunctionValue = math.max(fit.raw).val;
+        recentMinFunctionValue = math.min(fit.raw).val;
+        updateDistribution();
+    }
+    
+//    private IntDouble[] computePenalties() {
+//    	int i, j, iNk;
+//    	/* penalize repairment, eg. for boundaries */
+//    	// TODO: figure out whether the change of penalty is too large or fast which can disturb selection
+//    	//       this depence in particular on the length of fit.medianDeltaFit
+//    	if (true || countiter < fit.deltaFitHist.length || countiter % 1*(N+2) == 0) {
+//    		// minimum of differences with distance lambda/2, better the 25%tile?
+//    		// assumes sorted array!! 
+//    		int ii = (sp.getLambda()) / 2;
+//    		double medianDeltaFit = Math.abs(fit.funValues[ii].d - fit.funValues[0].d);
+//    		for (i = 1; i + ii < sp.getLambda(); ++i) 
+//    			// minimum because of outliers 
+//    			medianDeltaFit = Math.min(medianDeltaFit, Math.abs(fit.funValues[ii+i].d - fit.funValues[i].d));
+//    		medianDeltaFit /= sigma * sigma; // should be somehow constant, because dfit depends on sigma (verified on sphere)
+//    		if (medianDeltaFit > 0) {
+////  			System.out.println("set" + medianDeltaFit + " " + math.median(fit.medianDeltaFit));
+//    			if (fit.idxDeltaFitHist == -1) // first time: write all fields
+//    				for (i = 0; i < fit.deltaFitHist.length; ++i)
+//    					fit.deltaFitHist[i] = medianDeltaFit;
+//    			if (++fit.idxDeltaFitHist == fit.deltaFitHist.length)
+//    				fit.idxDeltaFitHist = 0;
+//    			// save last five values in fit.medianDeltaFit
+//    			fit.deltaFitHist[fit.idxDeltaFitHist] = medianDeltaFit;
+//    		}                
+//    	}
+//    	/* calculate fitness by adding function value and repair penalty */
+//    	double penfactor = 1. * 5. * math.median(fit.deltaFitHist);
+//    	for (iNk = 0; iNk < sp.getLambda(); ++iNk) {
+//    		double sqrnorm = 0;
+//    		double prod = Math.pow(math.prod(diagD), 1.0/(double)N);
+//    		/* calculate C^-1-norm of Delta x: norm(D^(-1) * B^(-1) * (Delta x))^2 */
+//    		for (i = 0; i < N; ++i) {
+//    			double sum = 0.0;
+//    			for (j = 0, sum = 0.; j < N; ++j)
+//    				sum += B[j][i] * ((arxrepaired[fit.funValues[iNk].i][j] - arx[fit.funValues[iNk].i][j]));
+//    			sqrnorm += math.square(sum / (Math.pow(diagD[i], 0.9) * Math.pow(prod, 0.10))); // regularization to I
+//    		}
+//    		// sqrnorm/N equals approximately 1/sigma^2
+//    		fit.fitness[iNk].d = fit.funValues[iNk].d + penfactor * sqrnorm / (N+2); // / (sigma * sigma);
+//    		fit.fitness[iNk].i = fit.funValues[iNk].i;
+//    		// System.out.println(math.median(fit.medianDeltaFit) + " " + sqrnorm / (N+2)); // / (sigma * sigma));
+//    	}
+////  	if (countiter % 10 == 1)
+////  	System.out.println(math.median(fit.medianDeltaFit) + " " + sqrnorm);
+//    	return fit.fitness;
+//
+//    }
+
+    private void updateDistribution() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1728-1853 into a separate method.